routing.py

```diff
@@ -0,0 +1,272 @@
+        """
+
+        self.templates_dir = templates_dir
+
+    def get_languages(self, preferred_order=[]):
+        """
+        Find the available languages that exist as template folders
+        """
+
+        available_languages = []
+        top_folders = os.listdir(self.templates_dir)
+
+        for folder in top_folders:
+            if is_language(folder):
+                available_languages.append(folder)
+
+        weighted_languages = []
+
+        if available_languages and preferred_order:
+            for item in available_languages:
+                try:
+                    weight = preferred_order.index(item)
+                except ValueError:
+                    weight = available_languages.index(item) + len(
+                        preferred_order
+                    )
+
+                weighted_languages.append(
+                    (weight, item)
+                )
+
+            weighted_languages = sorted(weighted_languages)
+
+            available_languages = [item[1] for item in weighted_languages]
+
+        return available_languages
+
+    def try_alternate_path(self, path):
+        """
+        If the path is for a file, see if the directory exists instead.
+        If the path is for a directory, see if the file exists instead.
+        If the alternate exists, return it. Otherwise return None.
+        """
+
+        # Try the other URL form
+        if path.endswith('/'):
+            new_path = path.rstrip('/')
+            if os.path.isfile(self.templates_dir + get_file(new_path)):
+                return new_path
+        else:
+            new_path = path + '/'
+            if os.path.isfile(self.templates_dir + get_file(new_path)):
+                return new_path
+
+    def try_language_version_path(self, path, languages, versions):
+        """
+        Given a URL path that doesn't contain any language or version
+        information, and a list of possible languages and versions,
+        try adding each language and version to the URL until we find
+        an existing file, and then return the path for that file
+        """
+
+        # Build up language and version URLs
+        search_paths = []
+
+        if languages:
+            for language in languages:
+                search_paths.append("/" + language + path)
+
+        if versions:
+            new_paths = []
+            for version in versions:
+                search_paths.append("/" + version + path)
+
+                for search_path in search_paths:
+                    new_paths.append("/" + version + search_path)
+
+            search_paths += new_paths
+
+        for path in search_paths:
+            alt_path = path + '/'
+            if alt_path.endswith('//'):
+                alt_path = alt_path[:-2]
+
+            template = self.templates_dir + get_file(path)
+            alt_template = self.templates_dir + get_file(alt_path)
+
+            if os.path.isfile(template):
```

Comment:
I don't really like "breaking" a `for loop` with a return.
I think here there is a problem, can't you put the `if statement` outside of the loop? Otherwise this means the array `search_paths` is not constructed well

Reply:
Why don't you like breaking a for loop? I think it's pretty clear in this case. And not that it will probably matter all that much, but it's actually more efficient. Every new iteration of the loop does a new filesystem look-up, so we should really break out of the loop as soon as possible.

I don't understand your point about the array not being constructed - the array is not modified during this for loop. It's already complete before this loop starts.

Reply:
okay I understand we can get in efficiency by doing this. I just think it is always harder to maintain a code like this. 

(my bad for the `search_paths` I understand better what it is doing)

Reply:
Okay. I understand the concern - return statements in the middle of functions can be confusing. But in this case, the return is still at the end of the function, and as the point of the function is to search for stuff I think this is quite simple to understand.

I'd rather not have hard-and-fast rules in things like this and instead always look for examples of things that are written in overly complex ways.

In this case I think this is the simplest way to write this function.

Reply:
I'm a fan of returning a function at its earliest point and try to where I can.
To me, it makes the logic more explicit. If it gets to this point, it exits the function. Otherwise you need to scan ahead and check what happens.

I think returning early actually makes the code more maintainable and easier to read. If this is ever not the case then it could be split up more, or it can be flexible and return later. As Robin put it, I like it not being a hard and fast rule.